### PR TITLE
feat(gossipsub): support version 1.2.0

### PR DESCRIPTION
**Pull request by diegomrsantos**
- Support GossipSub 1.2 in the code
- Add utility template `template exclIfIt*[T](set: var HashSet[T], condition: untyped)`
- Renaming

closes https://github.com/vacp2p/nim-libp2p/issues/1102

### CODE REVIEW

**Comment by diegomrsantos** [comment]
This PR depends on this being fixed https://github.com/vacp2p/nim-libp2p/issues/1115

**Comment by diegomrsantos** [comment]
- Send `idontwant` only to peers that support GossipSub 1.2
- Add utility template `template exclIfIt*[T](set: var HashSet[T], condition: untyped)`

Those are gonna be moved to a new PR cause of https://github.com/vacp2p/nim-libp2p/issues/1115.

**Comment by diegomrsantos** [comment]
> I'm not an expert on GS by any means, but I'm expecting something like `if GossipSubCodec_12 in g.codecs: # do something GS1.2 related`. From what I understand, we just add the codecs (which needed to be done) but there's no difference between GossipSubCodec_10, GossipSubCodec_11 and GossipSubCodec_12.

I believe `g.codecs` are the versions supported by the current code, instead of the one negotiated with the peer - `pubsubpeer.codec`. Behaving like 1.2 - sending `idontwant` - will be implemented in a new PR as mentioned here: https://github.com/vacp2p/nim-libp2p/pull/1106#issuecomment-2149411500. 

It seems that our current code can negotiate 1.0 but will still behave like 1.1, as there's no check on how the protocol should behave depending on the version negotiated with a remote peer.

File: tests/pubsub/utils.nim [hunk 1]
```diff
@@ -26,9 +26,9 @@ type
 
 proc getPubSubPeer*(p: TestGossipSub, peerId: PeerId): PubSubPeer =
   proc getConn(): Future[Connection] =
-    p.switch.dial(peerId, GossipSubCodec)
+    p.switch.dial(peerId, GossipSubCodec_12)
```
Comment (AlejandroCabeza) [inline tests/pubsub/utils.nim:29]: What happens if dialing 1.2.0 fails? Should we attempt to dial with a lower version?

Comment (diegomrsantos) [reply tests/pubsub/utils.nim:29]: This is only for tests. Production code tries all versions.

**Review by lchenut — APPROVED** [review]

**Review by lchenut — APPROVED** [review]

**Pull request merged**